### PR TITLE
Update error message for error 6651 (`Data location must be ...`)

### PR DESCRIPTION
**Pull request by jhb10c**
External and public functions are not allowed to have parameters whose data location is storage. The error message has been updated to reflect this fact. 

This was originally requested in  #[11987](https://github.com/ethereum/solidity/issues/11987)

### CODE REVIEW

**Comment by cameel** [comment]
Your PR does not actually change the the code that generates the message. It only changes the tests. You need to apply the change here:
https://github.com/ethereum/solidity/blob/d10e668f4f89878983fcaa46e7b88462ffa4380c/libsolidity/analysis/DeclarationTypeChecker.cpp#L370-L391

**Comment by jhb10c** [comment]
That makes sense. 

I have added in the line below "string(_variable.isExternalCallableParameter() ? " external" : "")": string(_variable.isPublicCallableParameter() ? " public" : "") and updated the tests using isoltest. This time it should work.

File: libsolidity/analysis/DeclarationTypeChecker.cpp [hunk 1]
```diff
@@ -383,6 +383,7 @@ void DeclarationTypeChecker::endVisit(VariableDeclaration const& _variable)
 					string(_variable.isReturnParameter() ? "return " : "") +
 					"parameter in" +
 					string(_variable.isExternalCallableParameter() ? " external" : "") +
+                    string(_variable.isPublicCallableParameter() ? " public" : "") +
```
Comment (jhb10c) [inline libsolidity/analysis/DeclarationTypeChecker.cpp]: This adds the word public in front function for error message 6651.

Comment (cameel) [inline libsolidity/analysis/DeclarationTypeChecker.cpp]: Your line is indented with spaces while the rest of the code is using tabs. This is probably why [`chk_coding_style`](https://circleci.com/gh/ethereum/solidity/860718) job is failing.

**Review by tied2token — DISMISSED** [review]
It didn't seem to error for me

**Comment by jhb10c** [comment]
@tied2token, I used spaces instead of tabs in the c++ code. This has been updated.

File: libsolidity/analysis/DeclarationTypeChecker.cpp [hunk 2]
```diff
@@ -383,6 +383,7 @@ void DeclarationTypeChecker::endVisit(VariableDeclaration const& _variable)
 					string(_variable.isReturnParameter() ? "return " : "") +
 					"parameter in" +
 					string(_variable.isExternalCallableParameter() ? " external" : "") +
+					string(_variable.isPublicCallableParameter() ? " public" : "") +
```
Comment (cameel) [inline libsolidity/analysis/DeclarationTypeChecker.cpp]: This is correct but I think we should also be doing this for return parameters. I see a lot of messages in tests that still say "in function" and those are either for storage return parameters or for cases where a parameter has no location at all.

Actually, I see no reason not to make it simple and just always include the visibility in the message. You can get visibility like this (not tested but should work):
```c++
string visibility;
if (auto const* callable = dynamic_cast<CallableDeclaration const*>(_variable.scope()))
    visibility = Declaration::visibilityToString(callable->visibility());
```

**Review by cameel — COMMENTED** [review]
Looks good already but we could do more. See my comment below.

Also, please squash your commits so that each one represents a single logical change (with no separate commits for PR fixes).

File: libsolidity/analysis/DeclarationTypeChecker.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/analysis/DeclarationTypeChecker.cpp]: Just a heads up: If you ask for the visibility of a constructor, you will get a failing assertion.

Comment (cameel) [reply libsolidity/analysis/DeclarationTypeChecker.cpp]: ok, yeah, this may need some exta ifs.

Comment (jhb10c) [reply libsolidity/analysis/DeclarationTypeChecker.cpp]: Gotcha, @cameel. The first part seems straight forward: update the message for storage return parameters and parameters where type is not defined by including the function type. 

An example of the second type would be like "Data location must be "storage", "memory" or "calldata" for return parameter in function, but none was given." . The function type should be mentioned in the message. 

Thanks for the heads up @chriseth. To check that the CallableDeclaration object is not a constructor, it should look something like:

`string visibility;
if (auto const* callable = dynamic_cast<FunctionDeclaration const*>(_variable.scope()))
{if (!*callable.isConstructor())
       visibility = Declaration::visibilityToString(callable->visibility());
.....
}`

Or would it be better to leave the type as a CallableDeclaration and check that if it is a FunctionDeclaration then check if it is a constructor or not?

Comment (cameel) [reply libsolidity/analysis/DeclarationTypeChecker.cpp]: Actually, I see that there's already a check for a constructor parameter in the branch above (`if (_variable.isConstructorParameter())`) so I think you do not have to worry about this case. It's already handled.

File: libsolidity/analysis/DeclarationTypeChecker.cpp [hunk 3]
```diff
@@ -378,12 +378,17 @@ void DeclarationTypeChecker::endVisit(VariableDeclaration const& _variable)
...
+							string visibility;
+							if (auto const* callable = dynamic_cast<CallableDeclaration const*>(_variable.scope()))
+								visibility = Declaration::visibilityToString(callable->visibility());
+							errorString +=
+							" for " +
+							string(_variable.isReturnParameter() ? "return " : "") +
+							"parameter in" +
+							string(visibility.empty() ? visibility : " " + visibility) +
+							" function";
+					}
```
Comment (cameel) [inline libsolidity/analysis/DeclarationTypeChecker.cpp]:
```suggestion
				string visibility;
				if (auto const* callable = dynamic_cast<CallableDeclaration const*>(_variable.scope()))
					visibility = Declaration::visibilityToString(callable->visibility());
				errorString +=
					" for " +
					string(_variable.isReturnParameter() ? "return " : "") +
					"parameter in" +
					string(visibility.empty() ? visibility : " " + visibility) +
					" function";
			}
```

File: libsolidity/analysis/DeclarationTypeChecker.cpp [hunk 4]
```diff
@@ -378,12 +378,17 @@ void DeclarationTypeChecker::endVisit(VariableDeclaration const& _variable)
...
-					" function";
+			{
+							string visibility;
+							if (auto const* callable = dynamic_cast<CallableDeclaration const*>(_variable.scope()))
+								visibility = Declaration::visibilityToString(callable->visibility());
+							errorString +=
+							" for " +
+							string(_variable.isReturnParameter() ? "return " : "") +
+							"parameter in" +
+							string(visibility.empty() ? visibility : " " + visibility) +
```
Comment (cameel) [inline libsolidity/analysis/DeclarationTypeChecker.cpp]: Since we're touching this code, we could also change the string conversion to a more concise one:
```suggestion
							_variable.isReturnParameter() ? "return "s : ""s +
							"parameter in" +
							visibility.empty() ? visibility : " "s + visibility +
```

File: libsolidity/analysis/DeclarationTypeChecker.cpp [hunk 5]
```diff
@@ -378,12 +378,17 @@ void DeclarationTypeChecker::endVisit(VariableDeclaration const& _variable)
...
+							" function";
```
Comment (cameel) [inline libsolidity/analysis/DeclarationTypeChecker.cpp]: We could use our `joinHumanReadable()` to make this more concise (needs `#include <libsolutil/StringUtils.h`). This removes the need to worry about including or omitting the spaces between components:
```suggestion
							string parameterKind = _variable.isReturnParameter() ? "return parameter"s : "parameter"s;
							errorString += " " + joinHumanReadable({"for", parameterKind, "in", visibility, "function"}, " ");
```

**Review by cameel — COMMENTED** [review]
Overall the code seems correct. Only style needs some tweaking.

I see one more case where this message does not include visibility: arguments function pointers (see `function_type_argument_external.sol`, `function_type_return_external.sol` and `data_location_in_function_type_fail.sol`). We should fix them too but it's also fine to merge what we have here and do it in a separate PR in my opinion.

Please rebase on latest `develop` - it has fixes for the CI jobs that are failing in your PR. And we need to see that this passes all of them.

File: libsolidity/analysis/DeclarationTypeChecker.cpp (hunk 5 above)
Comment (jhb10c) [reply libsolidity/analysis/DeclarationTypeChecker.cpp]: I tried using joinHumanReadable() but it could not find the function even with the header included. I have opted to for the second approach. 

There are still whitespace issues that I am fat fingering. ( I am missing how to obtain the correct styling completely....)

Comment (cameel) [reply libsolidity/analysis/DeclarationTypeChecker.cpp]: Maybe that's because this file does not use the namespace where the function is defined yet. Please try `util::joinHumanReadable()` or add `using namespace solidity::util;` at the top of the file.

Comment (cameel) [reply libsolidity/analysis/DeclarationTypeChecker.cpp]: The whitespace issue seems to be that you're indenting with spaces while we use tabs for .cpp files.

I see that the `chk_coding_style` job detects this but does not print any message about what is wrong. We should change that at some point.

**Comment by leonardoalt** [comment]
Coding style still needs some fixing.

File: libsolidity/analysis/DeclarationTypeChecker.cpp (hunk 5 above)
Comment (jhb10c) [reply libsolidity/analysis/DeclarationTypeChecker.cpp]: got you, let me try again real fast

**Comment by jhb10c** [comment]
I'm a tech boomer. The whitespace on my IDE was set to space and not tab....

File: libsolidity/analysis/DeclarationTypeChecker.cpp [hunk 6]
```diff
@@ -376,12 +376,17 @@ void DeclarationTypeChecker::endVisit(VariableDeclaration const& _variable)
 			if (_variable.isConstructorParameter())
 				errorString += " for constructor parameter";
 			else if (_variable.isCallableOrCatchParameter())
+			{
+				string visibility;
```
Comment (chriseth) [inline libsolidity/analysis/DeclarationTypeChecker.cpp:380]: Maybe
```suggestion
				string functionVisibility;
```
?

Comment (jhb10c) [reply libsolidity/analysis/DeclarationTypeChecker.cpp:380]: Gotcha. I will update this and submit later tonight.

**Comment by chriseth** [comment]
Closing due to inactivity. If you still want to work on in, feel free to re-open.
